### PR TITLE
The interpolation machinery is developed for a brick type  geometry. …

### PR DESCRIPTION
**Pull request by skandalaCLIMA**
…This works for the case where the element deformation/warping is of the type x = X(ξ1); y = Y(ξ2); z = Z(ξ3).

### CODE REVIEW

**Comment by jkozdon** [comment]
Thanks for starting to work on this. Glad someone is taking this up.

Just wanted to make sure you knew that we have some machinery in the mesh stuff to do [interpolation matrices] already. 

For the GPU, it would likely be a bit more efficient to use [barycentric weights] directly and apply everything in a matrix free manner. This would be more memory efficient on the GPU since you only need to load an array for weights as opposed to many matrices.

Since we have a tensor product structure, you can do dimension by dimension application of the algorithms right out of [Berrut & Trefethen]

[interpolation matrices]: https://github.com/climate-machine/CLIMA/blob/65fd895e9f8f36f17bcc914820ee99926231f76b/src/Mesh/Elements.jl#L87-L118
[barycentric weights]: https://github.com/climate-machine/CLIMA/blob/65fd895e9f8f36f17bcc914820ee99926231f76b/src/Mesh/Elements.jl#L27-L50
[Berrut & Trefethen]: https://doi.org/10.1137/S0036144502417715

File: src/Mesh/Interpolation.jl [hunk 1]
```diff
@@ -0,0 +1,371 @@
...
+            [ ξ[el][i] = 2.0 * ( xg[ xsten[1,dim] + i - 1] - xbndl[1,dim] ) / (xbndl[2,dim]-xbndl[1,dim]) -  1.0 for i in 1:n123[dim] ]
+        end
+
+        x[el] = zeros(FT,ndim,prod(n123))
+        V[el] = zeros(FT,prod(n123))
+
+        ctr = 1
+
+        for k in 1:n123[3], j in 1:n123[2], i in 1:n123[1]
+            x[el][1,ctr] = x1g[ xsten[1,1] + i - 1 ]
```
Comment (simonbyrne) [inline src/Mesh/Interpolation.jl]: it's not a huge deal, but this does add an extra layer of indirection. I would suggest defining
```julia
x_el = zeros(FT,ndim,prod(n123))
```
before the loop, do 
```suggestion
            x_el[1,ctr] = x1g[ xsten[1,1] + i - 1 ]
```
etc. in the loop, and
```julia
x[el] = x_el
```
outside the loop

File: src/Mesh/Interpolation.jl [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+            x[el][2,ctr] = x2g[ xsten[1,2] + j - 1 ]
+            x[el][3,ctr] = x3g[ xsten[1,3] + k - 1 ]
+            ctr += 1
+        end
+    end # el loop
+    #-----------------------------------------------------------------------------------
+    return new{FT}(El, Nel, xbnd, xres, ξ1, ξ2, ξ3, x, V)
+
+end # function Interpolation_Brick -> inner constructor
```
Comment (simonbyrne) [inline src/Mesh/Interpolation.jl]: if indented correctly, these comments should be necessary.

File: src/Mesh/Interpolation.jl [hunk 3]
```diff
@@ -0,0 +1,371 @@
...
+
+export Interpolation_Brick, interpolate_brick!, interpolate_brick_mf!, Interpolation_Cubed_Sphere, invert_trilear_mapping_hex, interpolate_cubed_sphere!
+
+#--------------------------------------------------------
+
+# This interpolation routine works for a brick, where stretching/compression happens only along the x, y & z axis.
+# Here x = X(ξ1); y = Y(ξ2); z = Z(ξ3)
+
+struct Interpolation_Brick{FT <:AbstractFloat}
+    El::UnitRange{Int64}
```
Comment (simonbyrne) [inline src/Mesh/Interpolation.jl]: Might be a good idea to use the same names as in Topologies
```suggestion
    realelems::UnitRange{Int64}
```

File: src/Mesh/Interpolation.jl [hunk 4]
```diff
@@ -0,0 +1,371 @@
...
+#--------------------------------------------------------
+end # struct Interpolation_Brick 
+#--------------------------------------------------------
+function interpolate_brick!(intrp_brck::Interpolation_Brick, sv::AbstractArray{FT}, st_no::T, poly_order::T) where {T <: Integer, FT <: AbstractFloat}
+
+    qm1 = poly_order + 1
+    m1_r, m1_w = GaussQuadrature.legendre(FT,qm1,GaussQuadrature.both)
+    wb = Elements.baryweights(m1_r)
+    #-----for each element elno 
+    for el in 1:intrp_brck.Nel
```
Comment (simonbyrne) [inline src/Mesh/Interpolation.jl]:
```suggestion
    for el in intrp_brck.realelems
```

File: src/Mesh/Interpolation.jl [hunk 5]
```diff
@@ -0,0 +1,371 @@
...
+#--------------------------------------------------------
+function interpolate_brick_mf!(intrp_brck::Interpolation_Brick, sv::AbstractArray{FT}, st_no::T, poly_order::T) where {T <: Integer, FT <: AbstractFloat}
+    qm1 = poly_order + 1
+    m1_r, m1_w = GaussQuadrature.legendre(FT,qm1,GaussQuadrature.both)
+    wb = Elements.baryweights(m1_r)
+    phir = Vector{FT}(undef,qm1)
+    phis = Vector{FT}(undef,qm1)
+    phit = Vector{FT}(undef,qm1)
+
+    for el in 1:intrp_brck.Nel #-----for each element elno
```
Comment (simonbyrne) [inline src/Mesh/Interpolation.jl]:
```suggestion
    for el in intrp_brck.realelems
```

File: src/Mesh/Interpolation.jl [hunk 6]
```diff
@@ -0,0 +1,371 @@
...
+    reorder = zeros(T, 6*nvert*nhor*nhor)
+
+    for i in 1:length(flip_ord), j in 1:nvert
+        reorder[j + (i-1)*nvert] = (flip_ord[i] - 1)*nvert + j
+    end
+
+    temp = invperm(grid.topology.origsendorder) 
+    ξ1, ξ2, ξ3        = map( i -> zeros(FT,i), zeros(T,Nel)), map( i -> zeros(FT,i), zeros(T,Nel)), map( i -> zeros(FT,i), zeros(T,Nel))
+    radc, latc, longc = map( i -> zeros(FT,i), zeros(T,Nel)), map( i -> zeros(FT,i), zeros(T,Nel)), map( i -> zeros(FT,i), zeros(T,Nel))
+    V                 = map( i -> zeros(FT,i), zeros(T,Nel))
```
Comment (simonbyrne) [inline src/Mesh/Interpolation.jl]:
```suggestion
```
Not needed with changes below.

File: src/Mesh/Interpolation.jl [hunk 7]
```diff
@@ -0,0 +1,472 @@
...
+                end
+                d = sum(phi)
+                phi ./= d 
+            end
+
+            intrp_cs.V[el][i] = 0.0
+            for ik in 1:qm1, ij in 1:qm1, ii in 1:qm1
+                intrp_cs.V[el][i] += lag[ii + (ij-1)*qm1 + (ik-1)*qm1*qm1] * 
+                                     phir[ii] * phis[ij] * phit[ik]
+            end
```
Comment (simonbyrne) [inline src/Mesh/Interpolation.jl]:
```suggestion
           W = 0.0
           @inbounds for ik in 1:qm1
                V = 0.0
                for ij in 1:qm1
                   U = 0.0
                   for ii in 1:qm1
                     U += lag[ii + (ij-1)*qm1 + (ik-1)*qm1*qm1] * phir[ii] 
                   end
                   V += phis[ij] * U
              end
              W += phis[ik] * V
        end
        intrp_cs.V[el][i] = W
```

File: src/Mesh/Interpolation.jl [hunk 8]
```diff
@@ -0,0 +1,472 @@
...
+
+            phir = Elements.interpolationmatrix(m1_r, intrp_brck.ξ1[el], wb)
+            phis = Elements.interpolationmatrix(m1_r, intrp_brck.ξ2[el], wb)
+
+            for k in 1:l3 # interpolating point-by-point
+                ξ3 = intrp_brck.ξ3[el][k]
+                interpolationmatrix_1pt(m1_r,ξ3,wb,phit)
+
+                for j in 1:l2, i in 1:l1
+                        intrp_brck.V[el][i + (j-1)*l1 + (k-1)*l1*l2] = 0.0
```
Comment (charleskawczynski) [inline src/Mesh/Interpolation.jl]:
```suggestion
                        intrp_brck.V[el][i + (j-1)*l1 + (k-1)*l1*l2] = FT(0)
```

File: src/Mesh/Interpolation.jl [hunk 9]
```diff
@@ -0,0 +1,472 @@
...
+
+    ξ1::Vector{Vector{FT}} # unique ξ1 coordinates of interpolation points within each element 
+    ξ2::Vector{Vector{FT}} # unique ξ2 coordinates of interpolation points within each element 
+    ξ3::Vector{Vector{FT}} # unique ξ3 coordinates of interpolation points within each element 
+  
+    x::Vector{Array{FT}}   # x[elno][3,npts] -> (x1,x2,x3) coordinates of portion of interpolation grid embedded within each element 
+
+    V::Vector{Vector{FT}}  # interpolated variable within each element
+#--------------------------------------------------------
+    function InterpolationBrick(grid::DiscontinuousSpectralElementGrid, xres, ::Type{FT}) where FT <: AbstractFloat
```
Comment (charleskawczynski) [inline src/Mesh/Interpolation.jl]:
```suggestion
    function InterpolationBrick(grid::DiscontinuousSpectralElementGrid{FT}, xres) where FT <: AbstractFloat
```
You'll need to change the way this is called, too.

File: src/Mesh/Interpolation.jl [hunk 10]
```diff
@@ -0,0 +1,472 @@
...
+    realelems::UnitRange{Int64}
+
+    xbnd::Array{FT}        # domain bounds, [2(min/max),ndim]
+    xres::Array{FT}        # resolutions in x1, x2, x3 directions for the uniform grid
+
+    ξ1::Vector{Vector{FT}} # unique ξ1 coordinates of interpolation points within each element 
+    ξ2::Vector{Vector{FT}} # unique ξ2 coordinates of interpolation points within each element 
+    ξ3::Vector{Vector{FT}} # unique ξ3 coordinates of interpolation points within each element 
+  
+    x::Vector{Array{FT}}   # x[elno][3,npts] -> (x1,x2,x3) coordinates of portion of interpolation grid embedded within each element
```
Comment (charleskawczynski) [inline src/Mesh/Interpolation.jl]:
```suggestion
    x::Vector{Array{FT,2}}   # x[elno][3,npts] -> (x1,x2,x3) coordinates of portion of interpolation grid embedded within each element 
```

Comment (charleskawczynski) [reply src/Mesh/Interpolation.jl]: Consider adding ranks to other arrays without rank specified

File: src/Mesh/Interpolation.jl [hunk 11]
```diff
@@ -0,0 +1,472 @@
...
+        for dim in 1:ndim 
+            xbnd[1,dim], xbnd[2,dim] = extrema(grid.topology.elemtocoord[dim,:,:])
+        end
+        x1g = range(xbnd[1,1], xbnd[2,1], step=xres[1])
+        x2g = range(xbnd[1,2], xbnd[2,2], step=xres[2])
+        x3g = range(xbnd[1,3], xbnd[2,3], step=xres[3]) 
+        #-----------------------------------------------------------------------------------
+        realelems =  grid.topology.realelems # Element (numbers) on the local processor
+        Nel       = length(realelems)
+        n123      = zeros(T,    ndim)     # # of unique ξ1, ξ2, ξ3 points in each cell
```
Comment (charleskawczynski) [inline src/Mesh/Interpolation.jl]: If these are not accumulated, you might want to use `Array{T}(undef, ...)`

File: test/Mesh/runtests.jl [hunk 12]
```diff
@@ -11,6 +11,8 @@ include("grid_integral.jl")
 include("filter.jl")
 include("Geometry.jl")
 
+include("interpolation.jl")
```
Comment (simonbyrne) [inline test/Mesh/runtests.jl]: since this uses MPI, it should be moved to the MPI jobs below

File: src/Mesh/Interpolation.jl [hunk 13]
```diff
@@ -0,0 +1,442 @@
...
+            x[el] = x_el
+        end # el loop
+        #-----------------------------------------------------------------------------------
+        return new{FT}(realelems, xbnd, xres, ξ1, ξ2, ξ3, x, V)
+    end
+#--------------------------------------------------------
+end # struct InterpolationBrick 
+#--------------------------------------------------------
+"""
+    function interpolate_brick!(intrp_brck::InterpolationBrick, sv::AbstractArray{FT}, st_no::T, poly_order::T) where {T <: Integer, FT <: AbstractFloat}
```
Comment (simonbyrne) [inline src/Mesh/Interpolation.jl]:
```suggestion
    interpolate_brick!(intrp_brck::InterpolationBrick, sv::AbstractArray{FT}, st_no::T, poly_order::T) where {T <: Integer, FT <: AbstractFloat}
```

File: src/Mesh/Interpolation.jl [hunk 14]
```diff
@@ -0,0 +1,442 @@
...
+function interpolate_cubed_sphere!(intrp_cs::InterpolationCubedSphere, sv::AbstractArray{FT}, st_no::T, poly_order::T) where {T <: Integer, FT <: AbstractFloat}
+    qm1 = poly_order + 1
+    Nel = length(intrp_cs.realelems)
+    m1_r, m1_w = GaussQuadrature.legendre(FT,qm1,GaussQuadrature.both)
+    wb = Elements.baryweights(m1_r)
+    phir = Vector{FT}(undef,qm1)
+    phis = Vector{FT}(undef,qm1)
+    phit = Vector{FT}(undef,qm1)
+
+    vout    = 0.0
```
Comment (simonbyrne) [inline src/Mesh/Interpolation.jl]:
```suggestion
    vout    = zero(FT)
```

File: src/Mesh/Interpolation.jl [hunk 15]
```diff
@@ -0,0 +1,442 @@
...
+        lat_min,   lat_max = FT(0.0), FT(π)                 # inclination/zeinth angle range
+        long_min, long_max = FT(0.0), FT(2*π)  			    # azimuthal angle range
+        rad_min,   rad_max = vert_range[1], vert_range[end] # radius range
+
+        realelems = grid.topology.realelems # Element (numbers) on the local processor
+        Nel = length(realelems)
+
+        Nel_glob = length(grid.topology.elems)
+        nvert = length(vert_range) - 1              # # of elements in vertical direction
+        nhor  = T( sqrt( Nel_glob / nvert / 6) )  # # of elements in horizontal direction
```
Comment (simonbyrne) [inline src/Mesh/Interpolation.jl]:
```suggestion
        nhor  = isqrt( div(Nel_glob, nvert * 6) )  # # of elements in horizontal direction
```

File: src/Mesh/Interpolation.jl [hunk 16]
```diff
@@ -0,0 +1,442 @@
...
+    ξ3::Vector{Vector{FT}} # ξ3 coordinates of interpolation points within each element 
+
+    radc::Vector{Vector{FT}}  # rad coordinates of interpolation points within each element
+    latc::Vector{Vector{FT}}  # lat coordinates of interpolation points within each element
+    longc::Vector{Vector{FT}} # long coordinates of interpolation points within each element
+
+    V::Vector{Vector{FT}}  # interpolated variable within each element
+  #--------------------------------------------------------
+    function InterpolationCubedSphere(grid::DiscontinuousSpectralElementGrid, vert_range::AbstractArray{FT}, lat_res::FT, long_res::FT, rad_res::FT) where {FT <: AbstractFloat}
+        T = Int
```
Comment (simonbyrne) [inline src/Mesh/Interpolation.jl]: Most uses of `T` below are unnecessary, or can be replaced by `Int`
```suggestion
```

File: src/Mesh/Interpolation.jl [hunk 17]
```diff
@@ -0,0 +1,442 @@
...
+        rad_min,   rad_max = vert_range[1], vert_range[end] # radius range
+
+        realelems = grid.topology.realelems # Element (numbers) on the local processor
+        Nel = length(realelems)
+
+        Nel_glob = length(grid.topology.elems)
+        nvert = length(vert_range) - 1              # # of elements in vertical direction
+        nhor  = T( sqrt( Nel_glob / nvert / 6) )  # # of elements in horizontal direction
+        nblck = nhor * nhor * nvert
+        Δh = 2.0 / nhor                             # horizontal grid spacing in unwarped grid
```
Comment (simonbyrne) [inline src/Mesh/Interpolation.jl]:
```suggestion
        Δh = 2 / nhor                             # horizontal grid spacing in unwarped grid
```

File: src/Mesh/Interpolation.jl [hunk 18]
```diff
@@ -0,0 +1,442 @@
...
+ - `st_no` # of state vector variable to be interpolated
+ - `poly_order` polynomial order for the simulation
+"""
+function interpolate_brick!(intrp_brck::InterpolationBrick, sv::AbstractArray{FT}, st_no::T, poly_order::T) where {T <: Integer, FT <: AbstractFloat}
+    qm1 = poly_order + 1
+    Nel = length(intrp_brck.realelems)
+    m1_r, m1_w = GaussQuadrature.legendre(FT,qm1,GaussQuadrature.both)
+    wb = Elements.baryweights(m1_r)
+
+    vout    = 0.0
```
Comment (simonbyrne) [inline src/Mesh/Interpolation.jl]:
```suggestion
    vout    = zero(FT)
```
here and below

File: src/Mesh/Interpolation.jl [hunk 19]
```diff
@@ -0,0 +1,442 @@
...
+            LinearAlgebra.BLAS.gemm!('T', 'T', alpha, reshape(lag,si, sj*sk), phir, beta, h1)
+            LinearAlgebra.BLAS.gemm!('T', 'T', alpha, reshape(h1, sj, sk*sr), phis, beta, h2)
+            LinearAlgebra.BLAS.gemm!('T', 'T', alpha, reshape(h2, sk, sr*ss), phit, beta, reshape(intrp_brck.V[el], sr*ss, st))
+            #---------------------------------------------------------------
+        end
+  #--------------------
+  end
+end
+#--------------------------------------------------------
+@inline function interpolationmatrix_1pt(rsrc::Vector{FT}, rdst::FT,
```
Comment (simonbyrne) [inline src/Mesh/Interpolation.jl]: it doesn't return a matrix, so maybe `interpolation_weights`?

Also would be good to add docs.

File: src/Mesh/Interpolation.jl [hunk 20]
```diff
@@ -0,0 +1,454 @@
...
+This interpolation function works for a brick, where stretching/compression happens only along the x, y & z axis.
+Here x1 = X1(ξ1); x2 = X2(ξ2); x3 = X3(ξ3)
+
+# input
+ - `intrp_brck` Initialized InterpolationBrick structure
+ - `sv` State vector
+ - `st_idx` # of state vector variable to be interpolated
+ - `poly_order` polynomial order for the simulation
+"""
+function interpolate_brick!(intrp_brck::InterpolationBrick, sv::AbstractArray{FT}, st_idx::T, poly_order::T) where {T <: Integer, FT <: AbstractFloat}
```
Comment (simonbyrne) [inline src/Mesh/Interpolation.jl]: we should make `poly_order` a field of the `InterpolationBrick` struct, since it is defined by the `grid`.

**Pull request merged**